File: python/symmetricalTree.py
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        
        
    def PreOrder(self, array: list) -> list:
        array.append(self.value)
        
        if self.left != None:
            array = self.left.PreOrder(array)
        if self.right != None:
            array = self.right.PreOrder(array)
        
        return array
    
    
    def InversePreOrder(self, array: list) -> list:
        array.append(self.value)
        if self.right != None:
            array = self.right.InversePreOrder(array)
        if self.left != None:
            array = self.left.InversePreOrder(array)
        
        return array
    
    
    
def PopulateBT() -> BinaryTree:
    tree = BinaryTree(1)
    tree.left = BinaryTree(2)
    tree.right = BinaryTree(2)
    
    left = tree.left
    right = tree.right
    
    left.left = BinaryTree(3)
    left.right = BinaryTree(4)
    
    right.left = BinaryTree(4)
    right.right = BinaryTree(3)
    
    return tree
    

# Time -> 0(N) Space -> 0(N)    
def Sol1() -> bool:
    arr1, arr2 = [], []
    
    tree = PopulateBT()
    if tree.left != None:
        arr1 = tree.left.PreOrder([])
    if tree.right != None:
        arr2 = tree.right.InversePreOrder([])
        
    return arr1 == arr2
    

# Time -> 0(N) Space -> 0(h) where h is the max height of the binary tree
def Sol2() -> bool:
    left_stack, right_stack = [], []
    
    tree = PopulateBT()
    left_stack.append(tree.left)
    right_stack.append(tree.right)
    
    while left_stack and right_stack:
        left_tree = left_stack.pop()
        right_tree = right_stack.pop()
        if left_tree == None and right_tree == None:
            continue
        
        if left_tree is None or right_tree is None or left_tree.value != right_tree.value:
            return False
        left_stack.append(left_tree.left)
        left_stack.append(left_tree.right)
        right_stack.append(right_tree.right)
        right_stack.append(right_tree.left)
            
            
    if left_stack or right_stack:
        return False
    
    return True    
        
        
    

print(Sol1())
print(Sol2())

File: go/symmetricalTree/symmetricalTree.go
package main


import (
	"fmt"
	"reflect"
)


type BinaryTree struct{
	Value int

	Left *BinaryTree
	Right *BinaryTree
}


func (tree *BinaryTree) PreOrder(array []int) []int {
	if tree == nil {return array}

	array = append(array, tree.Value)
	array = tree.Left.PreOrder(array)
	array = tree.Right.PreOrder(array)

	return array
}


func (tree *BinaryTree) InversePreOrder(array []int) []int {
	if tree == nil {return array}

	array = append(array, tree.Value)
	array = tree.Right.InversePreOrder(array)
	array = tree.Left.InversePreOrder(array)

	return array
}


func PopulateTree() *BinaryTree{
	tree := &BinaryTree{Value: 1}
	tree.Left = &BinaryTree{Value: 2}
	tree.Right = &BinaryTree{Value: 2}
	left := tree.Left
	right := tree.Right
	left.Left = &BinaryTree{Value: 3}
	left.Right = &BinaryTree{Value: 4}
	right.Left = &BinaryTree{Value: 4}
	right.Right = &BinaryTree{Value: 3}
	
	return tree
}


// Time -> 0(N) Space -> 0(N)
func Sol1() bool {
	tree := PopulateTree()

	var arr1, arr2 []int

	if tree.Left != nil {
		arr1 = tree.Left.PreOrder([]int{})
	}
	if tree.Right != nil {
		arr2 = tree.Right.InversePreOrder([]int{})
	}

	return reflect.DeepEqual(arr1, arr2)
}


// Time -> 0(N) Space -> 0(h) where h is the max height of the tree
func Sol2() bool {
	tree := PopulateTree()

	var leftStack, rightStack []*BinaryTree

	leftStack = append(leftStack, tree.Left)
	rightStack = append(rightStack, tree.Right)

	for len(leftStack) > 0 {
		left := leftStack[len(leftStack)-1]
		right := rightStack[len(leftStack)-1]

		leftStack = leftStack[:len(leftStack)-1]
		rightStack = rightStack[:len(rightStack)-1]

		if left == nil && right == nil {continue}

		if left == nil || right == nil || left.Value != right.Value {return false}

		leftStack = append(leftStack, left.Left)
		leftStack = append(leftStack, left.Right)
		rightStack = append(rightStack, right.Right)
		rightStack = append(rightStack, right.Left)
	}

	return true
}


func main() {
	fmt.Println(Sol1())
	fmt.Println(Sol2())
}